parsers/rwcpopmelody_parser.py

```diff
@@ -0,0 +1,197 @@
+import time
+import midi
+import pretty_midi
+import pandas as pd
+
+import jams
+
+
+def fill_file_metadata(jam, metadata, n_track):
+    """Fills the global metada into the JAMS jam."""
+    jam.file_metadata.artist = metadata['Artist (Vocal)'][n_track]
+    jam.file_metadata.title = metadata['Title'][n_track]
+
+    d_str = metadata['Length'][n_track]
+    jam.file_metadata.duration = (float(d_str.split(":")[0]) * 60 +
+                                  float(d_str.split(":")[1]))
+
+    # Store remaining RWC metadata in the sandbox
+    sandbox_dict = {
+        'Piece No.': metadata['Piece No.'][n_track],
+        'Cat. Suffix': metadata['Cat. Suffix'][n_track],
+        'Tr. No.': metadata['Tr. No.'][n_track],
+        'Singer Information': metadata['Singer Information'][n_track],
+        'Tempo': metadata['Tempo'][n_track],
+        'Live Instruments Used': metadata['Live Instruments Used'][n_track],
+        'Drum Information': metadata['Drum Information'][n_track]}
+
+    jam.sandbox.update(**sandbox_dict)
+
+
+def fill_annotation_metadata(annot):
+    """Fills the annotation metadata."""
+    annot.annotation_metadata.corpus = "RWC Music Database: Popular Music"
+    annot.annotation_metadata.version = "1.0"
+    annot.annotation_metadata.annotation_tools = ""
+    annot.annotation_metadata.annotation_rules = ""
+    annot.annotation_metadata.validation = ""
+    annot.annotation_metadata.data_source = ""
+    annot.annotation_metadata.curator = jams.Curator(name="Masataka Goto",
+                                                     email="m.goto@aist.go.jp")
+    annot.annotation_metadata.annotator = {}
+
+
+def create_jams(smf_file, out_file, metadata):
+    """
+    Creates a JAMS file from an RWC POP smf file (RM-P*.MID).
+    Note: only the notes of the MELODY track are kept!
+    """
+
+    # Load midi file
+    m = midi.read_midifile(smf_file)
+
+    # This will store the relevant MIDI tracks
+    melody = []
+
+    # Track 0 is metadata that we need
+    melody.append(m[0])
+
+    # Collect track text (convert to lower case and remove spaces)
+    track_text = ["".join(e.text.lower().split()) for t in m
+                  for e in t if isinstance(e, midi.TrackNameEvent)]
+
+    assert len(track_text) == len(m)
+
+    # Find the melody track: it should START with "melo", or if that's not
+    # there then "voca" (for vocal).
+    track_text = [t[:4] for t in track_text]
+    if 'melo' in track_text:
+        index = track_text.index('melo')
+    elif 'voca' in track_text:
+        index = track_text.index('voca')
+    else:
+        print("ABORTING TRACK: couldn't find melody for: %s" % smf_file)
+        return 0
+
+    melody.append(m[index])
+
+    # Create new midi pattern with just these tracks
+    m = midi.Pattern(tracks=melody, resolution=m.resolution,
+                     format=m.format, tick_relative=m.tick_relative)
+
+    # Write temporary midi file
+    temp_midi = out_file.replace(".jams", ".temp.mid")
```

Comment:
Can we improve this hack?  Preferably via one of the following:

- Use a properly secure temporary filename
- Skip tempfiles altogether and use a StringIO/BytesIO.  I'm not sure if midi and pretty_midi support that (but they should).

Finally, If any of those asserts below fail, then this will leave temp files on disk.  A better pattern here would be to wrap this all in a `try` block, and put the tempfile cleanup in a `finally` clause.  That way, even if the asserts fail, the cleanup still executes.